File: src/Responder.php
<?php

declare(strict_types=1);

namespace Bot;

use Bot\Dao\DaoInterface;
use Bot\Handlers\Factory;

final class Responder
{
	public function __construct(
		private DaoInterface $dao
	) {}

	public function respond(object $request): ?Response
	{
		$presentation = $this->findPresentation($request->message->from->id);
		$handler = Factory::create($presentation->status);
		$response = $handler->handle($request, $presentation);
		$this->dao->persist($presentation);

		return $response;
	}

	private function findPresentation(int $userId): Presentation
	{
		return $this->dao->find($userId) ?? new Presentation($userId);
	}
}

File: responder.go
package main

type Responder struct {
	dao     Dao
	handler Handler
}

type Response struct {
	text       string
	forceReply bool
}

func newResponder(dao Dao) *Responder {
	responder := &Responder{dao: dao}
	return responder
}

func (r *Responder) respond(u *TelegramUpdate) *Response {
	presentation := r.dao.find(u.Message.From.Id)
	r.handler = createHandler(presentation.Status)
	response := r.handler.generateResponse(u, presentation)
	r.handler.updatePresentation(u, presentation)
	r.dao.persist(presentation)
	return response
}

func (r *Response) toTelegramResponse(u *TelegramUpdate) *TelegramResponse {
	if r == nil {
		return nil
	}
	telegramResponse := TelegramResponse{
		Method:           "sendMessage",
		ChatId:           u.Message.Chat.Id,
		Text:             r.text,
		ReplyToMessageId: u.Message.Id,
	}
	if r.forceReply {
		telegramResponse.ReplyMarkup = struct {
			ForceReply bool `json:"force_reply"`
			Selective  bool `json:"selective"`
		}{true, true}
	}
	return &telegramResponse
}
